### ISSUE TITLE
Watchdog reset spin appeared on Raspberry Pi 4 in Debian Bullseye, but not present in Buster

### ISSUE THREAD

**Issue by steverpalmer** [systemd/systemd#21949]
**systemd version the issue has been seen with**
 > 247.3-6

**Used distribution**
 > Debian Bullseye and Raspberry Pi OS Bullseye

**Linux kernel version used** (`uname -a`)

 > `Linux rpi4-20210823 5.10.0-8-arm64 #1 SMP Debian 5.10.46-4 (2021-08-03) aarch64 GNU/Linux`

**CPU architecture issue was seen on**
 > ARMv8 on a Raspberry Pi 4

**Expected behaviour you didn't see**
 > n/a

**Unexpected behaviour you saw**
 > After enabling the systemd control of the Watchdog (setting RuntimeWatchdogSec=20), on reboot, the Pi booted to a login request, then automatically rebooted after about 3 seconds, and then kept rebooting.

**Steps to reproduce the problem**
 > Take any Bullseye release (Either Raspberry Pi OS or Debian's own Tested or Daily Build image) and flash on to Micro-SD card.  After first boot, edit the /etc/systemd/system.conf file to add "RuntimeWatchdogSec=20".  On reboot, the Pi goes into a reboot spin.  This does not happen with Buster release (systemd version 241)  (again either Raspberry Pi OS or Debian's own Daily Build image).

**Additional program output to the terminal or log subsystem illustrating the issue**

```text
Sorry, but I've found it difficult to get debug information when the device keeps resetting on me.
```

**Comment by bluca** [comment]
do you see anything useful in the console if you boot with systemd.log_level=debug on the kernel command line?

**Comment by steverpalmer** [comment]
So I edited the /boot/cmdline.txt file to read:

> console=serial0,115200 console=tty1 root=PARTUUID=1bac2e18-02 rootfstype=ext4 fsck.repair=yes rootwait systemd.log_level=debug

which I hope will cover the request, and did the "RuntimeWatchdogSec=20" edit to system.conf, and reboot...

The boot message scroll past fast, but I let it spin three times looking for anything distinctive.  All I can say for sure is that those line with the [...] markers all reported [OK] and nothing else such as [ERROR].

So I inspected the journalctl file for earlier boots and found:

> Dec 31 13:11:54 raspberrypi systemd[1]: Using hardware watchdog 'Broadcom BCM2835 Watchdog timer', version 0, device /dev/watchdog
> Dec 31 13:11:54 raspberrypi systemd[1]: Failed to set timeout to 600s: Invalid argument

Other parts of the journal reported:

> Dec 31 13:13:49 raspberrypi systemd[1]: Using hardware watchdog 'Broadcom BCM2835 Watchdog timer', version 0, device /dev/watchdog
> Dec 31 13:13:49 raspberrypi systemd[1]: Failed to set timeout to 20s: Invalid argument
> Dec 31 13:13:49 raspberrypi systemd[1]: Queued start job for default target Graphical Interface.
> Dec 31 13:13:49 raspberrypi systemd[1]: Failed to set timeout to 20s: Invalid argument
> Dec 31 13:13:49 raspberrypi systemd[1]: Failed to set timeout to 20s: Invalid argument
> Dec 31 13:13:49 raspberrypi systemd[1]: Failed to set timeout to 20s: Invalid argument

and

> Dec 31 13:27:16 raspberrypi systemd[1]: Failed to set timeout to 20s: Invalid argument
> -- Boot d6e05cd6f0194c37b7cdd0bfa94e2b5e --
> Dec 31 13:27:16 raspberrypi systemd[1]: systemd-update-utmp.service: starting held back, waiting for: systemd-tmpfiles-setup.service
> -- Boot 1f9e0e7b94bf44478c6b70d6ee090dcb --
> Dec 31 13:27:16 raspberrypi systemd[1]: Failed to set timeout to 20s: Invalid argument
> -- Boot d6e05cd6f0194c37b7cdd0bfa94e2b5e --
> Dec 31 13:27:16 raspberrypi systemd[1]: systemd-journald.service: Got notification message from PID 135 (FDSTORE=1)
> -- Boot 1f9e0e7b94bf44478c6b70d6ee090dcb --
> Dec 31 13:27:16 raspberrypi systemd[1]: Failed to set timeout to 20s: Invalid argument
> -- Boot d6e05cd6f0194c37b7cdd0bfa94e2b5e --
> Dec 31 13:27:16 raspberrypi systemd[1]: systemd-journald.service: Added fd 45 (n/a) to fd store.
> -- Boot 1f9e0e7b94bf44478c6b70d6ee090dcb --
> Dec 31 13:27:16 raspberrypi systemd[1]: Failed to set timeout to 20s: Invalid argument
> -- Boot d6e05cd6f0194c37b7cdd0bfa94e2b5e --
> Dec 31 13:27:16 raspberrypi systemd[1]: systemd-journald.service: Got notification message from PID 135 (FDSTORE=1)
> -- Boot 1f9e0e7b94bf44478c6b70d6ee090dcb --
> Dec 31 13:27:16 raspberrypi systemd[1]: Failed to set timeout to 20s: Invalid argument

I'm guessing, but the 20s and 600s could related to the `RuntimeWatchdogSec=20` and default `RebootWatchdogSec/ShutdownWatchdogSec=10min`?

**Comment by bluca** [comment]
Yes, that's the ioctl to configure the hardware watchdog failing:

https://github.com/systemd/systemd/blob/v247/src/shared/watchdog.c#L41

Are you sure the hardware supports those values? It's returning `EINVAL`

**Comment by steverpalmer** [comment]
Ahh! It is difficult to track down, but I found a 2019 post that mentions the maximum time allowed is 15 seconds.  I've used 20 seconds before and somehow "got away with it".  I've just tried changing `RuntimeWatchdogSec=14` and it seems to work.  Specifically, it boots OK without any spin, and then does do a reboot when I use a fork bomb.  It looks like my problem was the 20 seconds that I took verbatim from [here](http://0pointer.de/blog/projects/watchdog.html).  Doh!

(I suppose that the problem may also occur in some form for the `RebootWatchdogSec/ShutdownWatchdogSec` if it defaults to 10min??)

Thanks Luca and Happy New Year!
